scene/resources/style_box.cpp

```diff
@@ -316,67 +333,239 @@ bool StyleBoxFlat::get_border_blend() const {
+	float offset[4] = { 0.5, 0.75, 0, 0.25 };
+	float factor = offset[corner_index];
+	float rad = corner_radius;
+	Point2 from = Point2(cosf(factor * 2 * Math_PI) * rad, sinf(factor * 2 * Math_PI) * rad) + pos;
+	for (int i = 0; i <= vert_count / 4; i++) {
+		factor = offset[corner_index] + ((float)i / (float)vert_count);
+		Point2 to = Point2(cosf(factor * 2 * Math_PI) * rad, sinf(factor * 2 * Math_PI) * rad) + pos;
+		vs->canvas_item_add_line(p_canvas_item, from, to, col, 1, true);
+		from = to;
+	}
+}
+inline void draw_rounded_rect(VisualServer *vs, RID p_canvas_item, Rect2 rect, const int const_corner_radius[4], Color col_top, Color col_bottom, Color col_left, Color col_right, int border, bool filled) {
+
+	int corner_radius[4] = { const_corner_radius[0], const_corner_radius[1], const_corner_radius[2], const_corner_radius[3] };
+	for (int i = 0; i < 4; i++) {
+		if (corner_radius[i] * 2 > rect.size.height)
+			corner_radius[i] = rect.size.height / 2;
+		if (corner_radius[i] * 2 > rect.size.width)
+			corner_radius[i] = rect.size.width / 2;
 	}
 
-	if (draw_center)
-		vs->canvas_item_add_rect(p_canvas_item, Rect2(r.pos, r.size), bg_color);
+	Point2 tl = rect.pos + Vector2(corner_radius[0], corner_radius[0]);
+	Point2 tr = Point2(rect.pos.x + rect.size.x - corner_radius[1], rect.pos.y + corner_radius[1]);
+	Point2 br = rect.pos + rect.size - Vector2(corner_radius[2], corner_radius[2]);
+	Point2 bl = Point2(rect.pos.x + corner_radius[3], rect.pos.y + rect.size.y - corner_radius[3]);
 
-	Rect2i r_add = p_rect;
-	vs->canvas_item_add_rect(p_canvas_item, Rect2(Point2i(r_add.pos.x - additional_border_size[MARGIN_LEFT], r_add.pos.y - additional_border_size[MARGIN_TOP]), Size2(r_add.size.width + additional_border_size[MARGIN_LEFT] + additional_border_size[MARGIN_RIGHT], additional_border_size[MARGIN_TOP])), light_color);
-	vs->canvas_item_add_rect(p_canvas_item, Rect2(Point2i(r_add.pos.x - additional_border_size[MARGIN_LEFT], r_add.pos.y), Size2(additional_border_size[MARGIN_LEFT], r_add.size.height)), light_color);
-	vs->canvas_item_add_rect(p_canvas_item, Rect2(Point2i(r_add.pos.x + r_add.size.width, r_add.pos.y), Size2(additional_border_size[MARGIN_RIGHT], r_add.size.height)), dark_color);
-	vs->canvas_item_add_rect(p_canvas_item, Rect2(Point2i(r_add.pos.x - additional_border_size[MARGIN_LEFT], r_add.pos.y + r_add.size.height), Size2(r_add.size.width + additional_border_size[MARGIN_LEFT] + additional_border_size[MARGIN_RIGHT], additional_border_size[MARGIN_BOTTOM])), dark_color);
+	PoolVector2Array corners = PoolVector2Array();
+	corners.append(tl);
+	corners.append(tr);
+	corners.append(br);
+	corners.append(bl);
+
+	//draw corners
+	for (int i = 0; i < 4; i++) {
+
+		Color col = col_top;
+		if (i > 1)
+			col = col_bottom;
+
+		if (filled) {
+			//the circle has to be drawn twice, otherwise the circles for the corners are not visible
+			//super buggy ;)
+			vs->canvas_item_add_circle(p_canvas_item, corners[i], (float)corner_radius[i], col);
```

Comment:
filled arc is a really good idea!! also fixes some of the artefacts when drawing with transparency